stetho/src/main/java/com/facebook/stetho/inspector/elements/ShadowDOM.java

```diff
@@ -0,0 +1,364 @@
+/*
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.inspector.elements;
+
+import com.facebook.stetho.common.Accumulator;
+import com.facebook.stetho.common.ListUtil;
+import com.facebook.stetho.common.Util;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.IdentityHashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public final class ShadowDOM implements DOMView {
+  private final Object mRootElement;
+  private final IdentityHashMap<Object, ElementInfo> mElementToInfoMap = new IdentityHashMap<>();
+  private boolean mIsUpdating;
+
+  public ShadowDOM(Object rootElement) {
+    mRootElement = Util.throwIfNull(rootElement);
+  }
+
+  @Override
+  public Object getRootElement() {
+    return mRootElement;
+  }
+
+  @Override
+  public ElementInfo getElementInfo(Object element) {
+    return mElementToInfoMap.get(element);
+  }
+
+  public UpdateBuilder beginUpdate() {
+    if (mIsUpdating) {
+      throw new IllegalStateException();
+    }
+
+    mIsUpdating = true;
+
+    return new UpdateBuilderImpl();
+  }
+
+  public static interface UpdateBuilder {
```

Comment:
Strongly prefer treating IntelliJ as that flexibility (it can very trivially divorce concrete impls from their callers and create the appropriate interfaces).  It's much more powerful than you're assuming, I think...

Also, classes have real performance implications on Android and should always be avoided when developer tooling can be used to solve the problem.